File: 1997-next-palindrome-using-same-digits/1997-next-palindrome-using-same-digits.cpp
class Solution {
public:
    string nextPalindrome(string num) {
        int n = num.size();
        if (n <= 1) return "";  // Single-digit numbers cannot have a larger palindrome

        // Get the first half of the palindrome
        string firstHalf = num.substr(0, n / 2);
        
        // Find the next lexicographical permutation manually
        int i = firstHalf.size() - 2;
        
        // Step 1: Find the first decreasing element from the right
        while (i >= 0 && firstHalf[i] >= firstHalf[i + 1]) {
            i--;
        }
        
        // If no valid permutation is possible, return ""
        if (i < 0) return "";

        // Step 2: Find the rightmost element greater than firstHalf[i]
        int j = firstHalf.size() - 1;
        while (firstHalf[j] <= firstHalf[i]) {
            j--;
        }
        
        // Step 3: Swap elements at i and j
        swap(firstHalf[i], firstHalf[j]);

        // Step 4: Reverse everything after index i
        reverse(firstHalf.begin() + i + 1, firstHalf.end());

        // Construct the new palindrome
        string nextPal = firstHalf;
        if (n % 2 == 1) {  // If the original number has an odd length, keep the middle character
            nextPal += num[n / 2];
        }
        
        reverse(firstHalf.begin(), firstHalf.end()); // Mirror the first half to construct the palindrome
        nextPal += firstHalf;

        return nextPal;
    }
};

File: 1997-next-palindrome-using-same-digits/1997-next-palindrome-using-same-digits.java
class Solution {
    public static boolean findNextPermutation(List<Character> digits) {
        // Step 1: Find the first digit that is smaller than the digit after it
        int i = digits.size() - 2;
        while (i >= 0 && digits.get(i) >= digits.get(i + 1)) {
            i--;
        }
        if (i == -1) {
            return false;
        }

        // Step 2: Find the next largest digit to swap with digits[i]
        int j = digits.size() - 1;
        while (digits.get(j) <= digits.get(i)) {
            j--;
        }

        // Step 3: Swap and reverse the rest to get the smallest next permutation
        Collections.swap(digits, i, j);
        Collections.reverse(digits.subList(i + 1, digits.size()));
        return true;
    }
    public String nextPalindrome(String numStr) {
        int n = numStr.length();

        if (n == 1) {
            return "";
        }

        int halfLength = n / 2;
        List<Character> leftHalf = new ArrayList<>();
        for (int i = 0; i < halfLength; i++) {
            leftHalf.add(numStr.charAt(i));
        }

        // Step 1: Get the next permutation for the left half
        if (!findNextPermutation(leftHalf)) {
            return "";
        }

        // Step 2: Form the next palindrome by mirroring the left half
        StringBuilder nextPalindrome = new StringBuilder();
        for (char c : leftHalf) {
            nextPalindrome.append(c);
        }

        if (n % 2 == 0) {
            nextPalindrome.append(new StringBuilder(nextPalindrome).reverse());
        } else {
            nextPalindrome.append(numStr.charAt(halfLength));
            nextPalindrome.append(new StringBuilder(nextPalindrome.substring(0, halfLength)).reverse());
        }

        // Step 3: Ensure the result is larger than the original number
        if (nextPalindrome.toString().compareTo(numStr) > 0) {
            return nextPalindrome.toString();
        }
        return "";
    }
}
